File: backend/app/models/module.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Module(Base):
    __tablename__ = "modules"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    display_name: Mapped[str] = mapped_column(String(128), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    icon: Mapped[str] = mapped_column(String(32), default="folder")

    fields_schema: Mapped[dict] = mapped_column(JSONB, nullable=False)
    status_lifecycle: Mapped[list] = mapped_column(JSONB, default=list)

    alerts_config: Mapped[list] = mapped_column(JSONB, default=list)
    actions_config: Mapped[list] = mapped_column(JSONB, default=list)
    views_config: Mapped[list] = mapped_column(JSONB, default=list)
    last_action_runs: Mapped[dict] = mapped_column(JSONB, default=dict)

    is_system: Mapped[bool] = mapped_column(Boolean, default=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: frontend/src/types/module.ts
export type FieldType =
  | "string"
  | "text"
  | "integer"
  | "float"
  | "boolean"
  | "date"
  | "datetime"
  | "url"
  | "email"
  | "enum"
  | "tags";

export interface FieldDefinition {
  name: string;
  type: FieldType;
  required?: boolean;
  default?: string | number | boolean | null;
  values?: string[];
  description?: string;
}

export interface ViewConfig {
  name: string;
  type: "list" | "grouped" | "kanban" | "table" | "calendar";
  group_by?: string;
  sort_by?: string;
  filters?: Record<string, unknown>;
}

export interface ActionConfig {
  id: string;
  type: "web_search" | "internal_scan" | "enrichment";
  name: string;
  description: string;
  trigger: "scheduled" | "on_demand" | "on_event";
  frequency?: string;
  config: Record<string, unknown>;
  mcp_servers?: string[];
}

export interface Module {
  id: string;
  name: string;
  display_name: string;
  description: string;
  icon: string;
  fields_schema: FieldDefinition[];
  status_lifecycle: string[];
  alerts_config: unknown[];
  actions_config: ActionConfig[];
  views_config: ViewConfig[];
  is_system: boolean;
  sort_order: number;
  last_action_runs: Record<string, string>;
  created_at: string;
  updated_at: string;
  note_count: number;
}
